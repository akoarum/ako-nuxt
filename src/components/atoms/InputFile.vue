<template>
  <label class="inputFile">
    <input
      v-bind="$attrs"
      :name="name"
      :required="required"
      type="file"
      class="inputFile__input"
      @input="handleInput"
    />
    <span class="inputFile__btn">{{ label }}</span>
    <span class="inputFile__name">{{ fileName }}</span>
  </label>
</template>

<script>
export default {
  inheritAttrs: false,
  props: {
    name: { type: String, required: true },
    label: { type: String, default: 'アップロードする' },
    required: { type: Boolean }
  },
  data() {
    return {
      fileName: ''
    }
  },
  methods: {
    handleInput(event) {
      const file = event.target.files[0]

      if (!file) {
        this.$emit('input', null)
        this.fileName = ''
        return
      }

      this.$emit('input', file)
      this.fileName = file.name
    }
  }
}
</script>

<style lang="scss" scoped>
.inputFile {
  position: relative;
}
.inputFile__input {
  position: absolute;
  left: 0;
  top: 0;
  width: 5px;
  height: 5px;
  opacity: 0;
}
.inputFile__btn {
  display: inline-block;
  border-radius: 2px;
  background: $color-theme;
  line-height: 1;
  color: #fff;
  transition: background 0.3s $easing;

  @include media-min {
    padding: 6px 12px;
    font-size: $fontSize-pc;

    &:hover,
    &:focus {
      background: #5dc5e8;
    }
  }

  @include media-max {
    padding: 7px 10px;
    font-size: $fontSize-sp;

    &:hover,
    &:focus {
      background: #5dc5e8;
    }
  }
}
.inputFile__name {
  display: inline;
  margin-left: 6px;
  font-size: 1.4rem;
}
</style>
