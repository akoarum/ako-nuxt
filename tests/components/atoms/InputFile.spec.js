import { shallowMount } from '@vue/test-utils'
import InputFile from '~/components/atoms/InputFile'

describe('InputFile', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(InputFile, {
      propsData: {
        name: 'test'
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('handleInput', () => {
      it('fileに値があればfileNameに値を追加してemitする', () => {
        const event = {
          target: {
            files: {
              0: { name: 'test' }
            }
          }
        }
        wrapper.vm.handleInput(event)
        expect(wrapper.vm.fileName).toBe('test')
        expect(wrapper.emitted().input).toBeTruthy()
      })

      it('fileに値がなければfileNameを空にする', () => {
        const event = {
          target: {
            files: {}
          }
        }
        wrapper.vm.handleInput(event)
        expect(wrapper.vm.fileName).toBe('')
        expect(wrapper.emitted().input).toBeTruthy()
      })
    })
  })
})
