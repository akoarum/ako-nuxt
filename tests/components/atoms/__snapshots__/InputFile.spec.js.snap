// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFile initialized snapshot 1`] = `
<label
  class="inputFile"
>
  <input
    class="inputFile__input"
    name="test"
    type="file"
  />
   
  <span
    class="inputFile__btn"
  >
    アップロードする
  </span>
   
  <span
    class="inputFile__name"
  >
    
  </span>
</label>
`;
